src/CodeGen_D3D12Compute_Dev.cpp

```diff
@@ -0,0 +1,1154 @@
+    }
+
+    cache.clear();
+}
+
+void CodeGen_D3D12Compute_Dev::CodeGen_D3D12Compute_C::visit(const Select *op) {
+    ostringstream rhs;
+    string true_val = print_expr(op->true_value);
+    string false_val = print_expr(op->false_value);
+    string cond = print_expr(op->condition);
+    rhs << print_type(op->type)
+        << "("   << cond
+        << " ? " << true_val
+        << " : " << false_val
+        << ")";
+    print_assignment(op->type, rhs.str());
+}
+
+void CodeGen_D3D12Compute_Dev::CodeGen_D3D12Compute_C::visit(const Allocate *op) {
+
+    if (op->name == "__shared") {
+        // Already handled
+        op->body.accept(this);
+    } else {
+        open_scope();
+
+        debug(2) << "Allocate " << op->name << " on device\n";
+
+        debug(3) << "Pushing allocation called " << op->name << " onto the symbol table\n";
+
+        // Allocation is not a shared memory allocation, just make a local declaration.
+        // It must have a constant size.
+        int32_t size = op->constant_allocation_size();
+        user_assert(size > 0)
+            << "Allocation " << op->name << " has a dynamic size. "
+            << "Only fixed-size allocations are supported on the gpu. "
+            << "Try storing into shared memory instead.";
+
+        do_indent();
+        stream << print_storage_type(op->type) << ' '
+               << print_name(op->name) << "[" << size << "];\n";
+        do_indent();
+
+        Allocation alloc;
+        alloc.type = op->type;
+        allocations.push(op->name, alloc);
+
+        op->body.accept(this);
+
+        // Should have been freed internally
+        internal_assert(!allocations.contains(op->name));
+
+        close_scope("alloc " + print_name(op->name));
+    }
+}
+
+void CodeGen_D3D12Compute_Dev::CodeGen_D3D12Compute_C::visit(const Free *op) {
+    if (op->name == "__shared") {
+        return;
+    } else {
+        // Should have been freed internally
+        internal_assert(allocations.contains(op->name));
+        allocations.pop(op->name);
+        do_indent();
+    }
+}
+
+string CodeGen_D3D12Compute_Dev::CodeGen_D3D12Compute_C::print_assignment(Type type, const string& rhs)
+{
+    string rhs_modified = print_reinforced_cast(type, rhs);
+    return( CodeGen_C::print_assignment(type, rhs_modified) );
+}
+
+string CodeGen_D3D12Compute_Dev::CodeGen_D3D12Compute_C::print_vanilla_cast(Type type, string value_expr)
+{
+    ostringstream ss;
+    ss << print_type(type) << "(" << value_expr << ")";
+    return(ss.str());
+}
+
+string CodeGen_D3D12Compute_Dev::CodeGen_D3D12Compute_C::print_reinforced_cast(Type type, string value_expr)
+{
+    if (type.is_float())
+    {
+        return( print_vanilla_cast(type, value_expr) );
+    }
+
+    if (type.is_bool())
+    {
+        return( print_vanilla_cast(type, value_expr) );
+    }
+
+    if (type.bits() == 32)
+    {
+        return( print_reinterpret_cast(type, value_expr) );
+    }
+
+    // for signed types: shift-up then shift-down
```

Comment:
Fun stuff :) Please add in the comment that this is necessary because d3d12 only supports 32 bit ints